From e8d7df42fcf714660ddb3c4dfc8b1461068428ad Mon Sep 17 00:00:00 2001
From: Xianzhu Wang <wangxianzhu@chromium.org>
Date: Mon, 11 Nov 2024 18:36:53 +0000
Subject: [PATCH] [RasterInducingScroll] Fix crash when overflow clip is in layer state

The original code assumed that a DrawScrollingContentsOp must be after
the containing ConversionContext pushed a clip state, but that's not
always true. In very rare cases, the overflow clip corresponding to a
scroll translation may be in the layer state.

Bug: 377576484, 360635902
Change-Id: I32d85b6926420d45da4c46b38c4e40e8607b604c
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6002490
Reviewed-by: Philip Rogers <pdr@chromium.org>
Commit-Queue: Xianzhu Wang <wangxianzhu@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1381292}
---

--- a/third_party/blink/renderer/platform/graphics/compositing/paint_chunks_to_cc_layer.cc
+++ b/third_party/blink/renderer/platform/graphics/compositing/paint_chunks_to_cc_layer.cc
@@ -147,7 +147,7 @@
   ConversionContext(const PropertyTreeState& layer_state,
                     const gfx::Vector2dF& layer_offset,
                     Result& result,
-                    const StateEntry* outer_state_stack_top = nullptr)
+                    const HeapVector<StateEntry>* outer_state_stack = nullptr)
       : chunk_to_layer_mapper_(layer_state, layer_offset),
         current_transform_(&layer_state.Transform()),
         current_clip_(&layer_state.Clip()),
@@ -155,7 +155,7 @@
         current_scroll_translation_(
             &current_transform_->NearestScrollTranslationNode()),
         result_(result),
-        outer_state_stack_top_(outer_state_stack_top) {}
+        outer_state_stack_(outer_state_stack) {}
   ~ConversionContext();
 
  private:
@@ -352,9 +352,9 @@
 
   Result& result_;
 
-  // Points to the top of stack_stack_ of the outer ConversionContext that
-  // initiated the current ConversionContext in EmitDrawScrollingContentsOp().
-  const StateEntry* outer_state_stack_top_ = nullptr;
+  // Points to stack_stack_ of the outer ConversionContext that initiated the
+  // current ConversionContext in EmitDrawScrollingContentsOp().
+  const HeapVector<StateEntry>* outer_state_stack_ = nullptr;
 };
 
 template <typename Result>
@@ -454,11 +454,10 @@
         &target_clip.LowestCommonAncestor(*current_clip_).Unalias();
     const auto* clip = current_clip_;
     while (clip != lca_clip) {
-      if (!state_stack_.size() && outer_state_stack_top_) {
+      if (!state_stack_.size() && outer_state_stack_ &&
+          !outer_state_stack_->empty() && outer_state_stack_->back().IsClip()) {
         // We are ending a clip that is started from the outer
-        // ConversionContext. outer_state_stack_top_ should be always the
-        // overflow clip of the current scroll translation.
-        CHECK(outer_state_stack_top_->IsClip());
+        // ConversionContext.
         return {ScrollTranslationAction::kEnd};
       }
       if (!state_stack_.size() || !state_stack_.back().IsClip()) {
@@ -790,11 +789,9 @@
   while (state_stack_.size() && state_stack_.back().IsClip()) {
     EndClip();
   }
-  if (!state_stack_.size() && outer_state_stack_top_) {
-    // outer_state_stack_top_ should be always the overflow clip of the current
-    // scroll translation. The outer ConversionState should continue to end the
-    // clips.
-    CHECK(outer_state_stack_top_->IsClip());
+  if (!state_stack_.size() && outer_state_stack_ &&
+      !outer_state_stack_->empty() && outer_state_stack_->back().IsClip()) {
+    // The outer ConversionState should continue to end the clips.
     return {ScrollTranslationAction::kEnd};
   }
   return {};
@@ -896,7 +893,7 @@
   auto scrolling_contents_list = base::MakeRefCounted<cc::DisplayItemList>();
   ConversionContext<cc::DisplayItemList>(
       PropertyTreeState(scroll_translation, *current_clip_, *current_effect_),
-      gfx::Vector2dF(), *scrolling_contents_list, &state_stack_.back())
+      gfx::Vector2dF(), *scrolling_contents_list, &state_stack_)
       .Convert(chunk_it, end_chunk);
 
   EndTransform();
@@ -1031,7 +1028,7 @@
       continue;
     }
     if (action.type == ScrollTranslationAction::kEnd) {
-      if (outer_state_stack_top_) {
+      if (outer_state_stack_) {
         // Return to the calling EmitDrawScrollingContentsOp().
         return;
       } else {
--- a/third_party/blink/renderer/platform/graphics/compositing/paint_chunks_to_cc_layer_test.cc
+++ b/third_party/blink/renderer/platform/graphics/compositing/paint_chunks_to_cc_layer_test.cc
@@ -1275,6 +1275,73 @@
 }
 
 TEST_P(PaintChunksToCcLayerTest,
+       ScrollingContentsIntoDisplayItemListOverflowClipInLayerState) {
+  if (!RuntimeEnabledFeatures::RasterInducingScrollEnabled()) {
+    GTEST_SKIP();
+  }
+
+  auto scroll_state = CreateScrollTranslationState(
+      PropertyTreeState::Root(), -50, -60, gfx::Rect(5, 5, 20, 30),
+      gfx::Size(100, 200));
+  TestChunks chunks;
+  PropertyTreeState layer_state(t0(), scroll_state.Clip(), e0());
+  chunks.AddChunk(layer_state);
+  chunks.AddChunk(scroll_state);
+  chunks.AddChunk(layer_state);
+
+  auto cc_list = base::MakeRefCounted<cc::DisplayItemList>();
+  PaintChunksToCcLayer::ConvertInto(chunks.Build(), layer_state,
+                                    gfx::Vector2dF(), nullptr, *cc_list);
+
+  EXPECT_THAT(cc_list->paint_op_buffer(),
+              ElementsAre(PaintOpIs<cc::DrawRecordOp>(),  // chunk 0
+                          PaintOpIs<cc::DrawScrollingContentsOp>(),
+                          PaintOpIs<cc::DrawRecordOp>()));  // chunk 2
+  const auto& scrolling_contents_op =
+      static_cast<const cc::DrawScrollingContentsOp&>(
+          cc_list->paint_op_buffer().GetOpAtForTesting(1));
+  ASSERT_EQ(cc::PaintOpType::kDrawScrollingContents,
+            scrolling_contents_op.GetType());
+  EXPECT_THAT(scrolling_contents_op.display_item_list->paint_op_buffer(),
+              ElementsAre(PaintOpIs<cc::DrawRecordOp>()));  // chunk 1
+}
+
+TEST_P(PaintChunksToCcLayerTest,
+       ScrollingContentsIntoDisplayItemListOverflowClipInLayerStateWithEffect) {
+  if (!RuntimeEnabledFeatures::RasterInducingScrollEnabled()) {
+    GTEST_SKIP();
+  }
+
+  auto* effect = CreateOpacityEffect(e0(), t0(), nullptr, 0.5f);
+  auto scroll_state = CreateScrollTranslationState(
+      PropertyTreeState(t0(), c0(), *effect), -50, -60, gfx::Rect(5, 5, 20, 30),
+      gfx::Size(100, 200));
+  TestChunks chunks;
+  PropertyTreeState layer_state(t0(), scroll_state.Clip(), e0());
+  chunks.AddChunk(layer_state);
+  chunks.AddChunk(scroll_state);
+  chunks.AddChunk(layer_state);
+
+  auto cc_list = base::MakeRefCounted<cc::DisplayItemList>();
+  PaintChunksToCcLayer::ConvertInto(chunks.Build(), layer_state,
+                                    gfx::Vector2dF(), nullptr, *cc_list);
+
+  EXPECT_THAT(cc_list->paint_op_buffer(),
+              ElementsAre(PaintOpIs<cc::DrawRecordOp>(),      // chunk 0
+                          PaintOpIs<cc::SaveLayerAlphaOp>(),  // <effect>
+                          PaintOpIs<cc::DrawScrollingContentsOp>(),
+                          PaintOpIs<cc::RestoreOp>(),       // </effect>
+                          PaintOpIs<cc::DrawRecordOp>()));  // chunk 2
+  const auto& scrolling_contents_op =
+      static_cast<const cc::DrawScrollingContentsOp&>(
+          cc_list->paint_op_buffer().GetOpAtForTesting(2));
+  ASSERT_EQ(cc::PaintOpType::kDrawScrollingContents,
+            scrolling_contents_op.GetType());
+  EXPECT_THAT(scrolling_contents_op.display_item_list->paint_op_buffer(),
+              ElementsAre(PaintOpIs<cc::DrawRecordOp>()));  // chunk 1
+}
+
+TEST_P(PaintChunksToCcLayerTest,
        ScrollingContentsIntoDisplayItemListStartingFromNestedState) {
   if (!RuntimeEnabledFeatures::RasterInducingScrollEnabled()) {
     GTEST_SKIP();
--- a/third_party/blink/renderer/platform/testing/paint_property_test_helpers.h
+++ b/third_party/blink/renderer/platform/testing/paint_property_test_helpers.h
@@ -372,14 +372,13 @@
     CompositingReasons compositing_reasons = CompositingReason::kNone,
     MainThreadScrollingReasons main_thread_reasons =
         cc::MainThreadScrollingReason::kNotOpaqueForTextAndLCDText) {
-  PropertyTreeState state(PropertyTreeState::Root());
-  state.SetClip(*CreateClip(parent_state.Clip(), parent_state.Transform(),
-                            FloatRoundedRect(container_rect)));
-  state.SetTransform(*CreateScrollTranslation(
-      parent_state.Transform(), parent_scroll, offset_x, offset_y,
-      container_rect, contents_size, &state.Clip(), compositing_reasons,
-      main_thread_reasons));
-  return state;
+  auto* clip = CreateClip(parent_state.Clip(), parent_state.Transform(),
+                          FloatRoundedRect(container_rect));
+  auto* transform =
+      CreateScrollTranslation(parent_state.Transform(), parent_scroll, offset_x,
+                              offset_y, container_rect, contents_size, clip,
+                              compositing_reasons, main_thread_reasons);
+  return PropertyTreeState(*transform, *clip, parent_state.Effect());
 }
 
 inline PropertyTreeState CreateScrollTranslationState(
